### ISSUE TITLE
Windows: unlinkDir sometimes doesn't fire

### ISSUE THREAD

**Issue by silverwind** [paulmillr/chokidar#234]
I still have to explore if this is reproducible, but I just deleted a directory with a single child file using

````js
rimraf(directory, {maxBusyTries: 10});
````

and while I got the `unlink` event for the file, no `unlinkDir` was fired for the parent directory. It does seem to happen pretty rarely, I encountered it in maybe one of hundred deletions.

**Comment by silverwind** [comment]
Here's a test script I wrote for this:

````js
var fs = require('fs');
var rimraf = require('rimraf');
var stats = {runs:0}, ready;

function go() {
  console.log(stats);
  stats.runs++;
  fs.mkdir('test', function (err) {
    fs.writeFile('test/test.file', function () {
      rimraf('test', function (err) {
        setTimeout(go, 500);
      });
    });
  });
}

require('./').watch('.', {
  persistent: true
}).on('all', function(event) {
  if (!ready) return;
  if (!stats[event]) stats[event] = 0;
  stats[event]++;
}).on('ready', function () {
  ready = true;
  go();
});
````

And here's the result after 1000 runs:

````
{ runs: 1000,
  change: 1000,
  addDir: 1000,
  unlinkDir: 804,
  add: 757,
  unlink: 953 }
````

Looks quite bad with `unlinkDir` only firing 80% of the time. Will try to run on OS X too for comparison.

**Comment by es128** [comment]
It would probably improve if you upped the timeout.

Chokidar relies on `fs.watch` on Windows, which is well known to be far from perfect. Tough to do much about it if it totally misses events. In order to turn this into an action item, I'd need the raw events exposed alongside so that we can identify a pattern that can be mitigated for when this issue occurs.

**Comment by es128** [comment]
Also, spacing the fs operations a little more (wrap each callback layer in its own timeout) would likely improve the reliability.

**Comment by silverwind** [comment]
Yes, spacing out the fs changes to 250ms each certainly improves it:

````
{ runs: 100,
  change: 300,
  addDir: 100,
  add: 99,
  unlinkDir: 99,
  unlink: 100 }
````

And a second run: 

````
{ runs: 100,
  addDir: 100,
  add: 100,
  unlink: 100,
  unlinkDir: 100 }
````

It really is those back-to-back operations that are the issue here. Does Windows provide any native function that might be more precise?

**Comment by silverwind** [comment]
Also, if I activate just `usePolling` in the script above, with no spacing out write, the results look pretty strange, here's some prints:

````
{ runs: 1, addDir: 1, unlinkDir: 1 }
{ runs: 2, addDir: 1, unlinkDir: 1 }
{ runs: 3, addDir: 1, unlinkDir: 1 }
{ runs: 4, addDir: 1, unlinkDir: 1 }
{ runs: 5, addDir: 1, unlinkDir: 1 }
{ runs: 6, addDir: 1, unlinkDir: 1 }
{ runs: 7, addDir: 1, unlinkDir: 1 }
{ runs: 8, addDir: 1, unlinkDir: 1 }
{ runs: 9, addDir: 1, unlinkDir: 1 }
````

**Comment by es128** [comment]
Not sure what's going on in that last case.

The only idea that's coming to mind here is to proactively stat check a dir anytime all its children are removed in case `fs.watch` misses it.

**Comment by silverwind** [comment]
That sounds like a plan. In my use case, I only really care for `unlinkDir` event, the files itself don't really matter as they're gone anyways. Does chokidar already keep an in-memory list of all files inside each directory?

**Comment by es128** [comment]
> Does chokidar already keep an in-memory list of all files inside each directory?

Yes

**Comment by silverwind** [comment]
FYI: Playing a bit with the spacing, it seems at around 150ms is a point where I start to get all of the events.

````js
var fs = require('fs');
var rimraf = require('rimraf');
var stats = {runs:0}, ready;

var SPACING = 150;

function go() {
  console.log(stats);
  stats.runs++;
  fs.mkdir('test', function () {
    setTimeout(function() {
        fs.writeFile('test/test.file', function () {
            setTimeout(function () {
                rimraf('test', function () {
                    setTimeout(go, SPACING);
                });
            }, SPACING);
        });
    }, SPACING);
  });
}

require('./').watch('.', {
  persistent: true
}).on('all', function(event) {
  if (!ready) return;
  if (!stats[event]) stats[event] = 0;
  stats[event]++;
}).on('ready', function () {
  ready = true;
  go();
});
````

**Comment by es128** [comment]
@silverwind Could you please check whether 0c367f7 works to solve your issue?

**Comment by silverwind** [comment]
Looks much improved when runing the initial test:

````
{ runs: 1000,
  change: 1000,
  addDir: 995,
  unlinkDir: 995,
  add: 720,
  unlink: 720 }
````

The remaining five cases are probably more windows weirdness, but it's certainly a big improvement!

**Comment by silverwind** [comment]
On second thought, leaving it up to you to close this.

**Comment by es128** [comment]
Haha thanks for letting me do the honors. Yeah I'm going to merge the PR and close it. 

The underlying problem is that some events are missed when too many changes are happening at once, but the fs apis are too flawed to ever really reach perfection with this stuff.

If needing greater reliability than what you're seeing, I'd advise implementing another layer of double-checking the state of the file tree periodically, but that's something that would be too heavy to bake into chokidar.

**Comment by silverwind** [comment]
Yeah, in my case (an in-memory replication of a real file system tree), I came to the conclusion that fs watchers aren't a 100% thing due to OS bugs. I also noticed various kinds of ordering problems with the events, which lead me to do some really ugly hacks.

Luckily, my web clients initiate most changes so I will process these as primary truth source and only trust watchers when someone does direct changes on the file system, possibly backed by periodical full reads (https://github.com/silverwind/droppy/issues/234).
